man/dbemit.1
.TH DBEMIT 1 " 4 April 1996"
.FB
.SH NAME
dbemit \- emit a \fIFirstBase\fP database in comma separated format
.SH SYNOPSIS
.B dbemit
[\fB-b\fP]
[\fB-c\fP \fIchar_sep\fP] [\fB-f\fP] [\fB-q\fP] [\fB-s\fP] [\fIscreen\fP]
[\fB-t\fP] [\fB-v\fP] [\fB-y\fP] [\fB-B\fP] [\fB-\fP]
[\fB-d\fP \fIdatabase\fP] [\fB-i\fP \fIindex\fP]
.SH DESCRIPTION
.I Dbemit
emits all undeleted records of a \fIFirstBase\fP database in 
token separated format to an output file.
The set of processed records can be limited using the \fB-i\fP index specifier.
Each line contains one \fIFirstBase\fP record, with some token character
separating the fields. The default token is a comma.
.PP
The output file is the database name with a
.I .emit
extension. Creation of this output file can be overridden by using 
the \fB-\fP flag. This option will cause the output of the emit process to
go to standard output \(em most often the terminal.
.PP
Normally, 
\fIdbemit\fP(1) 
does not emit formula fields. However, using the 
.B -f
flag forces formula fields to be emitted along with real fields.
These formula emissions are calculated as they are being emitted.
.PP
The \fB-b\fP
flag will run the \fIFirstBase\fP tool in batch mode, suitable for shell
scripts.
This mode produces no output on the terminal, thus can also be used
when running the process in the background.
The \fB-y\fP flag will run the tool at the terminal without pausing to ask
the user if existing files can be destroyed and without pausing
for the users permission to continue.
.PP
The \fB-B\fP flag will enable the fastest throughput of
records from the database. If this flag can be used, the \fIdbemit\fP
generation is speeded up a great deal by internal record buffering.
Note that no index can be used with record buffering, and
that record locking is disabled.
.PP
\fIDbemit\fP(1)
normally prints most all data as it is stored in the database.
However, alpha fields will be quoted if the \fB-q\fP flag is used,
and date and dollar fields will be formatted using the \fB-v\fP flag.
Note that for alphanumeric
data, embedded quotes ('"') and backslashes (\\) are escaped using the
standard C and UNIX method of preceding the character with a 
backslash.
.PP
\fIDbemit\fP(1)
provides a convenient method of producing a 
database image file that can be
directly downloaded into 
a different database management system.
The 'comma separated field'
format is compatible with many other database
management packages.
.PP
It is also possible to change the field separator used by dbemit. 
This is done using the \fB-c\fP flag.
As an example, the following command will emit data from the
database \fIphonelist\fP using the index \fIfriends\fP.
.sp 1
dbemit -d phonelist -i friends - -c :
.sp 1
Furthermore, this data will be emitted to standard output, and the field
separator will be a colon, not a comma.
.PP
As another example, \fB-c \\|\fP could be used to change the separator to
the vertical bar character ('|'). (Note the escaping of the pipeline
character due to the shell.)
.PP
By using the \fB-t\fP flag,
\fIdbemit\fP(1) will produce output that is ready to be
submitted to \fItbl\fP(1), the table production program that works in
conjunction with \fInroff\fP(1). This means that the following command
would produce a formatted report from a database using non-database,
standard UNIX tools to do the formatting:
.sp 1
dbemit -d phonelist -t - | tbl | nroff -me -Tlp
.sp 1
Note that the entire phonelist is printed here \(em not just an indexed
sub-section of the database.
A database \fIscreen\fP mask can also be applied to the dbemit
process (the \fB-s\fP flag) to reorder or ignore particular fields.
.PP
It is also possible to upload values 
into a \fIFirstBase\fP database file. See
\fIdbload\fP(1).
.SH FILES
.PD 0
.TP 10
*.emit
extension for emitted databases.
.TP 10
dbase
default name of database.
.TP 10
*.cdb
\fIFirstBase\fP database.
.TP 10
*.map
record map of \fIFirstBase\fP database.
.TP 10
*.ddict
\fIFirstBase\fP database dictionary file.
.PD
.SH SEE ALSO
dbdbas(1), dbdscr(1), dbload(1)
.PP
.I FirstBase User's Guide and Reference Manual
.br


ChangeLog
2012-04-14    <john.bolding@epona.horseweb.com>

	* modified libscr/input.c -- read_kcheck -- to avoid a problem when:
	     EDITINPUT is OFF
	     two (2) ESCAPE's were hanging in read_kcheck

2011-09-29    <john.bolding@epona.horseweb.com>

	* added fflush(-1) to dbmacro via libmexpr/mac_file.c

2009-02-25    <john.bolding@epona.horseweb.com>

	* changed src/dbsql/tokenscript to allow for space between # and define.

2006-09-19  John  <john@cerebus.horseweb.com>

	* complete rewrite of libcc/subline.c to NOT write on the passed in s.

2003-03-29  John Bolding  <john@firstbase.com>

        * tweaked libmpars/tokenscript and dbsql/tokenscript
          to accomodate the bison output "# define" instead of "#define".

2001-09-01  John Bolding  <john@firstbase.com>

        * tweaked configure.in to not use termios.h if OS is sco3.2
	* src/libmexpr/mac_func.c
	* src/libmexpr/mac_str.c
	  augmented dbmacro to include
	     crypt(string, salt), returns an encrypted string ala crypt(3)

2001-03-19  John Bolding  <john@firstbase.com>

	* src/libmexpr/mac_sec.c: 
	* src/libmexpr/mac_func.c: 
	* src/libmexpr/mac_file.c: 
	* src/include/macro.h:
	  modified dbmacro functions as follows:
	  - modified mf_creat() to do a chmod on the file if it already exists.
	  - added an mf_mkstemp() that really calls mf_mktemp()
	    (since mf_mktemp really uses mkstemp(2), it closes the fd.
	     a real mf_mkstemp would modify the array passed in for
	     the user to keep. that did not happen this go around)
	  - renamed all the firstbase secure record functions to 
            r_owner() r_group() r_mode() r_chown() r_chgrp() r_chmod()
	  - added an mf_chmod() command

2001-03-18  John Bolding  <john@firstbase.com>

	* src/dbtools/dbmacro.c: fixed path_trans to be null before
	  its used. fixed to only look at extension of file name, .cgi,
	  to determine whether to go cgi mode. this is so that i can
	  test with program names like ndbmacro.cgi.

	* src/libcdb/cgi.c: modified cgi_read_get to FB_ERROR out
	  if the QUERY_STRING env variable is not set.

	* src/libcdb/System.c: seems linux system() call is a bit
	  flaky at times on some versions, so fb_system() has been
	  modified to not use the system() call at all.
	  now it uses execvp. in addition, changed to waitpid()
	  to try and get better response from cgi dbmacro scripts that
	  use the macro call system(). and, now pays attention to
	  status from waitpid() and exits if its strange.

2001-02-23  John Bolding  <john@firstbase.com>

	* src/libscr/screen.c: modified visible bell area to
	  use a ^G if its not defined.
	  this was a terminfo only bug --- termcap was already working.

2001-02-17  John Bolding  <john@firstbase.com>

	* release 9.3.1
	* man pages for section 4 were missing from distribution.
	  (finally noticed by <garak_dracul@hotmail.com>).

2001-02-16  John Bolding  <john@firstbase.com>

	* man.ps target now in man/Makefile

	* fixed use of mktemp and gets system wide

	* src/libinit/setup.c:
	  changed default behaviour. these are now default:
	     MENUPAUSE OFF
	     MAKEINDEX OFF
	     DECIMAL ON

	* quite a few bugs regarding #if clauses and FSYNC/FCNTL
	  these should have been HAVE_FSYNC or HAVE_FCNTL.
	  some were wrong. some were using FSYNC instead of HAVE_FSYNC.
	  this might have caused multiuser problems with records
	  stored across slow devices (NFS, etc).
	  all of these were introduced when firstbase went autoconf.
	
	* src/libcdb/System.c:
	  - added \" to list of characters that cause a complete system() call
	  - make return status better for those that fork and wait

	* a number of little bugs concering the full screen display
	  for devices bigger than 24 x 80.
	
	* 180 degree about face regarding TERMCAP vs TERMINFO:
	  since ncurses seems to be pretty common on Linux/FreeBSD/Etc,
	  the default is now to use the ncurses library and if this
	  is not found the HAVE_LIBNCURSES is undef'ed in
	  src/include/config.h.

	  Although I am not a TERMINFO fan, perhaps this will be good.
	  At some point, the firstbase screen libraries could
	  make better use of ncurses to optomize screen motion, etc.
	  Perhaps even support for colors, pop-up style ncurses items, etc.

	  So, TERMINFO via ncurses is the default.
	  And, TERMCAP is done only if necessary.

2001-02-05  John Bolding  <john@firstbase.com>

	* src/libscr/input.c: 
	* src/libscr/e_input.c: both of these were modified to handle
	  windows that change sizes while being used, ala SIGWINCH.

	* src/libscr/refresh.c: 
	* src/libscr/scrhdr.c: 
	* src/libscr/scrstat.c: 
	* src/libscr/scrtime.c: 
	* src/libinit/ender.c: 
	* src/libcdb/screrr.c:
	  all of these were modified to better handle standard
	  screen messages on surfaces bigger than 24 x 80,
	  and to restore headers and footers when the frame size changes.

	* src/libcdb/winsize.c: modified to handle different frame/window
	  sizes using an ioctl() call with TIOCGWINSZ or TIOCGSIZE,
	  then allowing environment variables LINES and COLUMNS to
	  override all else. (Thanks to <vanonim@dial.eunet.ch>).

	  finally, sanity checks are done to make sure
	  nothing interactive less than 24 x 80 gets through.

	* src/libinit/getargs.c: took out code that was getting in the way
	  of the new fb_winsize(). all sanity checks are now done at
	  screen setup time.

	* src/libscr/screen.c: force all columns and lines determinations
	  to be done inside of winsize, newly named fb_winsize().

2001-02-02  John Bolding  <john@firstbase.com>

	* src/libinit/setup.c: removed the LINES setup variable,
	     (also force makeindex to be OFF at startup).

2001-02-02  John Bolding  <john@firstbase.com>

	* .visual: changed to /usr/bin/vi (was /usr/ucb/vi).

	* configure.in (INCLUDES): made `ed' and one of `flex' or `lex'
	  and one of `bison', `byacc', or `yacc' required in order
	  to complete ./configure.

$Id: ChangeLog,v 9.9 2006/09/19 17:46:07 john Exp $


src/dbtools/dbemit.c
/*
 * Copyright (C) 2001 FirstBase Software, Inc. All rights reserved.
 *
 * $Id: dbemit.c,v 9.2 2002/03/09 00:07:35 john Exp $
 *
 * See the file LICENSE for conditions of use and distribution.
 *
 */

#ifndef LINT
static char Dbemit_sid[] = "@(#) $Id: dbemit.c,v 9.2 2002/03/09 00:07:35 john Exp $";
#endif

/*  
 *  dbemit.c - emit all the fb_field values of an entire database
 *    in comma seperated format .
 */

#include <fb.h>
#include <fb_vars.h>
#include <emit.h>

#if !FB_PROTOTYPES
static void demit();
static int emit();
static int semit();
static void finit();
static void begintable();
static void endtable();
void usage();
#else /* FB_PROTOTYPES */
extern main(int, char **);
static void demit(int, char **);
static int emit(fb_database *);
static int semit(fb_database *);
static void finit(char *);
static void begintable(void);
static void endtable(void);
void usage(void);
#endif /* FB_PROTOTYPES */

static fflag = 0;				/* 'force-formula-flag' */
static Bflag = 0;				/* Blocking flag */
static pflag = 0;				/* piping flag */
static stdflag = 0;				/* stdout flag */
static tableflag = 0;				/* for tbl output */
static quoteflag = 0;				/* quote flag for fields */
static verbose = 0;				/* for verbose dates */
static rflag = 0;				/* record flag */

static int fd;
static long n;
static char separator = ',';			/* standard separator */
fb_field **cdb_sp;				/* screen pointer */
short int cdb_sfields;				/* number of screen fields */

static char *S_QUOTE = "\"";
static char *S_BACKSLASH = "\\";

static char *USAGE = 
   "usage: dbemit [options] [-d dbase] [-i index] [-s screen]";

extern char *cdb_work_dir;
extern char *cdb_coname;

/*
 *  dbemit - main driver
 */
 
   main(argc, argv)
      int argc;
      char *argv[];
   
      {
	 int i;

         (void) Dbemit_sid;

         for (i = 1; i < argc; i++){
            if (argv[i][0] == CHAR_MINUS)
               if (argv[i][1] == NULL){
                  stdflag = 1;
                  cdb_batchmode = 1;
                  }
               else if (argv[i][2] == NULL){
                  switch (argv[i][1]) {
                     case 'v':
                        verbose = 1;
                        break;
                     case 't':
                        tableflag = 1;
                        separator = ':';
                        break;
                     case 'q':
                        quoteflag = 1;
                        break;
                     case 'c':
                        if (++i >= argc)
                           usage();
                        if (argv[i][1] != NULL)
                           usage();
                        separator = argv[i][0];
                        break;
                     case 'f':
                        fflag = 1;
                        break;
                     case 'B':
                        Bflag = 1;
                        break;
                     case 'p':
                        pflag = 1;
                        break;
                     case 'r':
                        rflag = 1;
                        break;
                     }
                  }
            }
         fb_getargs(argc, argv, FB_ALLOCDB);
	 
	 fb_allow_int();		/* allows instant piping */	
	 if (cdb_yesmode)
	    sleep(1);				/* some are too fast */
	 demit(argc, argv);
	 fb_ender();
      }

/*
 *  demit - emit loop for a single dbase.
 */

   static void demit(argc, argv)
      int argc;
      char *argv[];
      
      {
         char fname[FB_MAXNAME], msg[FB_MAXLINE];
	 fb_database *hp;
	 int i;
	 
	 hp = cdb_db;
	 if (fb_testargs(argc, argv, "-i") > 0)
	    fb_opendb(hp, READ, FB_WITHINDEX, FB_MUST_INDEX);
         else
	    fb_opendb(hp, READ, FB_NOINDEX, FB_MUST_INDEX);
         for (i = 0; i < hp->nfields; i++)
            fb_nounders(hp->kp[i]);
	 fb_gets_dict(argc, argv);
	 fb_scrhdr(hp, "Check Files"); fb_infoline();
	 fb_basename(fname, hp->dbase);
	 strcat(fname, EMITEXTENSION);
	 if (stdflag == 1)
	    fd = 1;	/* standard output */
	 else
            finit(fname);		/* global fd allows foreach */
         fb_w_init(1, fd, -1);
	 if (!cdb_batchmode && !cdb_yesmode){
	    sprintf(msg, "Ok to emit from Database %s (y/n)? ",
	       hp->dbase);
	    if (fb_mustbe('y', msg, cdb_t_lines, 1) == FB_ERROR)
	       fb_xerror(FB_ABORT_ERROR, SYSMSG[S_NOT_DONE], NIL);
	    }
	 if (tableflag)
	    begintable();
         if (emit(hp) == FB_ERROR)
            fb_serror(FB_MESSAGE, SYSMSG[S_NOT_DONE], NIL);
	 if (tableflag)
	    endtable();
	 fb_wflush(1);
	 if (fd != 1)
	    close(fd);
	 fb_closedb(hp);
      }

/* 
 *  emit - emit all fields of the fb_database 
 */
 
   static emit(hp)
      fb_database *hp;
      
      {
         n = 0L;
	 if (!cdb_batchmode){
	    FB_XWAIT();
	    fb_gcounter(n);
	    }
	 if (hp->dindex == NULL || hp->dindex[0] == NULL ||
               (hp->ifd < 0 && !hp->b_tree)){
            if (Bflag == 0)
	       return(fb_foreach(hp, semit));
            else
	       return(fb_blockeach(hp, semit));
            }
	 else
	    return(fb_forxeach(hp, semit));
      }

/* 
 *  semit - emit all the fields of a buffer 
 */

   static int semit(hp)
      fb_database *hp;
      
      {
         int i;
	 char *p, *df, rbuf[10];;
	 
	 n++;
	 if (!cdb_batchmode)
	    fb_gcounter(n);
         if (rflag){
            sprintf(rbuf, "%ld", hp->rec);
	    fb_nextwrite(0, rbuf);
	    fb_w_write(0, &separator);	/* output separator (,:) */
            }
         for (i = 0; i < cdb_sfields; i++)
	    if (fflag == 1 || 
	          (cdb_sp[i]->type != FB_FORMULA && cdb_sp[i]->dflink==NULL)){
	       if (i != 0)
		  fb_w_write(0, &separator);	/* output separator (,:) */
	       if (cdb_sp[i]->type == FB_FORMULA || cdb_sp[i]->dflink != NULL){
	          fb_fetch(cdb_sp[i], cdb_afld, hp);
		  df = cdb_afld;
		  }
	       else
	          df = cdb_sp[i]->fld;
               if (cdb_sp[i]->type == FB_BINARY)
                  ; /* do nothing */
	       else if (!(FB_OFNUMERIC(cdb_sp[i]->type)) && cdb_sp[i]->type != FB_FORMULA){
	          if (quoteflag)
	             fb_w_write(0, S_QUOTE);
		  if (cdb_sp[i]->type == FB_DATE && verbose){
		     strcpy(cdb_afld, df);
		     p = fb_formfield(cdb_bfld, cdb_afld, cdb_sp[i]->type, 8);
		     }
		  else
		     p = df;
		  for ( ; *p; p++){ /* escape emb quotes/backg */
		     if (*p == FB_NEWLINE)
		        fb_nextwrite(0, "\\n");
		     else {
		        if (*p == CHAR_QUOTE || *p == CHAR_BACKSLASH)
		           fb_w_write(0, S_BACKSLASH);
		        fb_w_write(0, p);
			}
		     }
	          if (quoteflag)
	             fb_w_write(0, S_QUOTE);
		  }
	       else {
	          if (cdb_sp[i]->type == FB_DOLLARS && verbose){
		     strcpy(cdb_afld, df);
		     fb_formdollar(cdb_bfld, cdb_afld, cdb_sp[i]->size);
                     fb_trim(cdb_bfld);
                     df = cdb_bfld;
		     }
		  fb_nextwrite(0, df);
		  }
	       }
	 fb_w_write(0, SYSMSG[S_STRING_NEWLINE]);
         if (pflag)
            fb_wflush(1);
	 return(FB_AOK);
      }

/* 
 *  finit - initialize --- merely check for existance. ask about overwrite 
 */
 
   static void finit(fname)
      char *fname;
      
      {
         char str[FB_MAXLINE];
	 
	 if (access(fname, 0) != -1 && !cdb_batchmode && !cdb_yesmode){
	    sprintf(str, 
	       "Permission to OVERWRITE %s (y = yes, <cr> = no)? ", fname);
	    if (fb_mustbe('y', str, 15, 10) == FB_ERROR)
	       fb_xerror(FB_ABORT_ERROR, SYSMSG[S_NOT_DONE], NIL);
	    }
	 sprintf(str, "to %s", fname);
	 if (!cdb_batchmode)
	    fb_scrstat(str);
         close(creat(fname, 0666));	/* create fname */
	 fd = fb_mustopen(fname, WRITE);
      }

/*
 * begintable - emit pre tbl output
 */

   static void begintable()
      {
         int i, j;
	 char name[FB_MAXNAME], buffer[FB_MAXLINE];
	 
	 sprintf (buffer, ".he '%s''Page %%'\n", cdb_coname);
	 fb_nextwrite(0, buffer);
	 sprintf (buffer, "Directory: %s\n.br\n", cdb_work_dir);
	 fb_nextwrite(0, buffer);
	 sprintf (buffer, "Database: %s\n.br\n",
            fb_basename(name, cdb_db->dbase));
	 fb_nextwrite(0, buffer);
	 if (cdb_db->ifd > 0){
	    sprintf (buffer, "Index: %s\n.br\n",
               fb_basename(name, cdb_db->dindex));
	    fb_nextwrite(0, buffer);
	    }
	 if (cdb_db->sdict != NIL){
	    sprintf (buffer, "Screen: %s\n.br\n",
               fb_basename(name, cdb_db->sdict));
	    fb_nextwrite(0, buffer);
	    }
	 sprintf(buffer, ".sp 2\n");
	 fb_nextwrite(0, buffer);
         sprintf(buffer, ".TS H\ntab (%c);\n", separator);
	 fb_nextwrite(0, buffer);
	 buffer[0] = NULL;
	 for (j = 1; j <= 2; j++){
	    for (i = 0; i < cdb_sfields; i++)
	       strcat(buffer, "c ");
	    strcat(buffer, "\n");
	    fb_nextwrite(0, buffer);
	    buffer[0] = NULL;
	    }
	 buffer[0] = NULL;
	 for (i = 0; i < cdb_sfields; i++)
	    switch (cdb_sp[i]->type){
	       case '$':
	       case 'f':
	       case 'F':
	       case 'n':
	       case 'N':
	       case 'c':
	          strcat(buffer, "n1 "); break;
	       default:
	          strcat(buffer, "l1 "); break;
	       }
	 strcat(buffer, ".\n");
	 fb_nextwrite(0, buffer);
	 for (i = 0; i < cdb_sfields; i++){
	    sprintf(buffer, "%s%c", cdb_sp[i]->id, separator);
	    fb_nextwrite(0, buffer);
	    }
	 fb_nextwrite(0, "\n");
	 for (i = 0; i < cdb_sfields; i++){
	    sprintf(buffer, "\\_%c", separator);
	    fb_nextwrite(0, buffer);
	    }
	 sprintf(buffer, "\n\n.TH\n");
	 fb_nextwrite(0, buffer);
      }

/*
 * endtable - emit end tbl output
 */

   static void endtable()
      {
         char buffer[FB_MAXLINE];
	 
         sprintf(buffer, ".TE\n");
	 fb_nextwrite(0, buffer);
      }

/* 
 * usage message 
 */

   void usage()
      {
         fb_xerror(FB_MESSAGE, USAGE, NIL);
      }
